Hessenberg decomposition

ks147 opened the pull request:
<!-- Hessenberg Decomposition -->



#### Implemented Upper Hessenberg Decomposition

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* matrices
  * Implemented Upper Hessenberg Decomposition for a matrix
<!-- END RELEASE NOTES -->

ks147 committed:
Implemented Upper Hessenberg decomposition for matrices

ks147 committed:
Added examples and some more tests

ks147 committed:
Removed newline to pass quality test

ks147 committed:
Removing import simplify to pass code quality test

ks147 commented:
@sylee957 @oscarbenjamin could you please review this PR.
Thanks

sylee957 commented:
The only commentary is why not naming `upper_hessenberg_decomposition`. We have old undecided problem about naming in #12389, but at least the convention is that longer named decompositions were named snakecase. (`jordan_decomposition, rank_decomposition ...`)

ks147 commented:
> The only commentary is why not naming `upper_hessenberg_decomposition`. We have old undecided problem about naming in #12389, but at least I would prefer that named decompositions should be named snakecase. (`jordan_decomposition, rank_decomposition ...`)

I'll change the name. I didn't know about the naming convention

ks147 committed:
Changed Function name according to sympy conventions

ks147 committed:
FOrgot to change function name in example in docs

sylee957 commented on sympy/matrices/decompositions.py:
```diff
@@ -1343,3 +1344,66 @@ def _QRdecomposition(M):
+def _upper_hessenberg_decomposition(A):
+    """Converts a matrix into Hessenberg matrix H
+
+        Returns 2 matrices H, P s.t.
+        $P H P^{T} = A$, where H is an upper hessenberg matrix
+        and P is an orthogonal matrix
+
+        Examples
+        ========
+        >>> from sympy import Matrix
+        >>> A = Matrix([
+        ...     [1,2,3],
+        ...     [-3,5,6],
+        ...     [4,-8,9],
+        ... ])
+        >>> H, P = A.upper_hessenberg_decomposition()
+        >>> H
+        Matrix([
+        [1,    6/5,    17/5],
+        [5, 213/25, -134/25],
+        [0, 216/25,  137/25]])
+        >>> P
+        Matrix([
+        [1,    0,   0],
+        [0, -3/5, 4/5],
+        [0,  4/5, 3/5]])
+        >>> P * H * P.H == A
+        True
+
+
+        References
+        ==========
+
+        .. https://mathworld.wolfram.com/HessenbergDecomposition.html
```
It's better to dedent these documentation one level

sylee957 replied:
```suggestion
    .. [#] https://mathworld.wolfram.com/HessenbergDecomposition.html
```

sylee957 commented on sympy/matrices/decompositions.py:
```diff
@@ -1343,3 +1344,66 @@ def _QRdecomposition(M):
+
+        Examples
+        ========
+        >>> from sympy import Matrix
```
```suggestion

    >>> from sympy import Matrix
```

ks147 committed:
Apply suggestions from code review

removing space in docstring

Co-authored-by: S.Y. Lee <sylee957@gmail.com>

sylee957 commented on sympy/matrices/decompositions.py:
```diff
@@ -1343,3 +1344,67 @@ def _QRdecomposition(M):
+def _upper_hessenberg_decomposition(A):
+    """Converts a matrix into Hessenberg matrix H
+
+        Returns 2 matrices H, P s.t.
+        $P H P^{T} = A$, where H is an upper hessenberg matrix
+        and P is an orthogonal matrix
+
+        Examples
+        ========
```
```suggestion
    Returns 2 matrices H, P s.t.
    $P H P^{T} = A$, where H is an upper hessenberg matrix
    and P is an orthogonal matrix

    Examples
    ========
```

sylee957 commented on sympy/matrices/decompositions.py:
```diff
@@ -1343,3 +1344,67 @@ def _QRdecomposition(M):
+        ========
+
+    >>> from sympy import Matrix
+        >>> A = Matrix([
+        ...     [1,2,3],
+        ...     [-3,5,6],
+        ...     [4,-8,9],
+        ... ])
+        >>> H, P = A.upper_hessenberg_decomposition()
+        >>> H
+        Matrix([
+        [1,    6/5,    17/5],
+        [5, 213/25, -134/25],
+        [0, 216/25,  137/25]])
+        >>> P
+        Matrix([
+        [1,    0,   0],
+        [0, -3/5, 4/5],
+        [0,  4/5, 3/5]])
+        >>> P * H * P.H == A
+        True
+
+
+        References
+        ==========
+
+    .. [#] https://mathworld.wolfram.com/HessenbergDecomposition.html
+    """
```
```suggestion
    >>> A = Matrix([
    ...     [1,2,3],
    ...     [-3,5,6],
    ...     [4,-8,9],
    ... ])
    >>> H, P = A.upper_hessenberg_decomposition()
    >>> H
    Matrix([
    [1,    6/5,    17/5],
    [5, 213/25, -134/25],
    [0, 216/25,  137/25]])
    >>> P
    Matrix([
    [1,    0,   0],
    [0, -3/5, 4/5],
    [0,  4/5, 3/5]])
    >>> P * H * P.H == A
    True


    References
    ==========

    .. [#] https://mathworld.wolfram.com/HessenbergDecomposition.html
    """
```

ks147 committed:
Dedenting Documentation

Co-authored-by: S.Y. Lee <sylee957@gmail.com>

sylee957 commented:
There are zero division corner cases in your method, try `Matrix([[1, 0, 0, 0], [0, 1, 0, 0], [1, 1, 0, 1], [1, 1, 1, 0]])`

ks147 committed:
Fixed issue of Matrix becoming NaN when a column is already upper Hessenberg

ks147 commented:
> There are zero division corner cases in your method, try `Matrix([[1, 0, 0, 0], [0, 1, 0, 0], [1, 1, 0, 1], [1, 1, 1, 0]])`

@sylee957 I've fixed this issue in the latest commit and I've added the given matrix as a test case in  test_decompositions.py

sylee957 commented on sympy/matrices/decompositions.py:
```diff
@@ -1343,3 +1344,72 @@ def _QRdecomposition(M):
+
+        u = H[j + 1:, j]
+
+        if u[1:, :] == M.zeros(1, u.rows - 1):
```
Why not use `is_zero_matrix`?

ks147 replied:
I didn't know about it, I was searching for something like that. I'll change it thanks
Thanks

sylee957 commented on sympy/matrices/decompositions.py:
```diff
@@ -1343,3 +1344,72 @@ def _QRdecomposition(M):
+    [1,    0,   0],
+    [0, -3/5, 4/5],
+    [0,  4/5, 3/5]])
+    >>> P * H * P.H == A
```
You also need to consider upgrading this method to work with complex matrix with complex householder transformation. Using `.H` for methods that work only with real matrices can be an abuse of notation.

ks147 replied:
```python
A = Matrix([
    [1, 1 + 2*I, 2],
    [1 - 3*I, -1, 3],
    [1 + 3*I, 1, 0]
])
H, P = A.upper_hessenberg_decomposition()
pprint(simplify(H))
```

    ⎡                                                        3⋅√2   8⋅√2⋅ⅈ        
    ⎢     1                                                  ──── - ──────        
    ⎢                                                         10      5           
    ⎢                                                                             
    ⎢              ⅈ⋅(-40⋅√5⋅(3 + 20⋅ⅈ) - 12⋅√10⋅(7 + 47⋅ⅈ) + 20⋅√5⋅(3 + 16⋅ⅈ)⋅(√2
    ⎢-√2 + 3⋅√2⋅ⅈ  ───────────────────────────────────────────────────────────────
    ⎢                                                  10⋅(√2 + 2)⋅(4⋅√5 + 3⋅√10) 
    ⎢                                                                             
    ⎢                   (4 - 3⋅ⅈ)⋅(146⋅√5⋅(√2 + 2) + 107⋅√10⋅(√2 + 2) - 10⋅√10⋅(23
    ⎢     0             ──────────────────────────────────────────────────────────
    ⎣                                                  50⋅(√2 + 2)⋅(4⋅√5 + 3⋅√10) 
    
                                                                                  
                                                                                  
                                                                                  
                                                                                  
     + 2) + √10⋅(42 + 227⋅ⅈ)⋅(√2 + 2))  (4 + 3⋅ⅈ)⋅(√10⋅(155 - 72⋅ⅈ)⋅(√2 + 2) + 6⋅√
    ──────────────────────────────────  ──────────────────────────────────────────
                                                                                  
                                                                                  
     + 12⋅ⅈ) - 12⋅√5⋅(27 + 14⋅ⅈ))               ⅈ⋅(-20⋅√10⋅(3 + 8⋅ⅈ) - 28⋅√5⋅(3 + 
    ─────────────────────────────               ──────────────────────────────────
                                                                                 1
    
         4⋅√2   11⋅√2⋅ⅈ                                                ⎤
         ──── + ───────                                                ⎥
          5        10                                                  ⎥
                                                                       ⎥
    5⋅(35 - 16⋅ⅈ)⋅(√2 + 2) + 4⋅√5⋅(-193 + 126⋅ⅈ) + 10⋅√10⋅(-55 + 36⋅ⅈ))⎥
    ───────────────────────────────────────────────────────────────────⎥
    50⋅(√2 + 2)⋅(4⋅√5 + 3⋅√10)                                         ⎥
                                                                       ⎥
    8⋅ⅈ) + √5⋅(6 + 31⋅ⅈ)⋅(√2 + 2) + 2⋅√10⋅(3 + 13⋅ⅈ)⋅(√2 + 2))         ⎥
    ──────────────────────────────────────────────────────────         ⎥
    0⋅(2⋅√10 + 3⋅√5)⋅(√2 + 2)                                          ⎦

ks147 committed:
Used is_zero_matrix instead of Matrix.zeros

sylee957 commented on sympy/matrices/decompositions.py:
```diff
@@ -1343,3 +1344,72 @@ def _QRdecomposition(M):
+    if not M.is_square:
+        raise NonSquareMatrixError("Matrix must be square.")
+
+    if M.is_upper_hessenberg:
+        return M, M.eye(M.cols)
```
```suggestion
```

Now, perhaps we can just drop this check because we can already detect if the matrix have hessenberg while running this.

ks147 replied:
Yeah we can drop this

ks147 committed:
Upgraded the method to work with complex matrices
